### ISSUE TITLE
Duplicate references when using the Importer

### ISSUE THREAD

**Issue by juanal98** [0xd4d/dnlib#342]
Good morning, I already commented a question #338 about how to add external methods to the assembly in .NET Core.

I have also read: #104 

I have adapted the AssemblyResolver so that it works correctly with .NET Core with the advice you gave me.

If I want to get for example, `System.Convert` resolving works correctly to get it from `System.Runtime.Extensions`, and everything works perfectly.

But for example, imagine that `System.Runtime.Extensions` is already in the module as a reference (because the original application already has it) there is the problem, instead of reusing it, add it again, and then the module has 2 identical references.

![Screenshot_10](https://user-images.githubusercontent.com/32484496/75624912-46048480-5bb9-11ea-9364-ab3c844fd4c8.png)

**This does not cause any problems, the application starts and works correctly even if there are two references the same, but of course, it is rare.**

So, I have done one thing, before looking for what library I need for "x" type, I check the current referenced libraries, and I try to solve the "x" type with each of them.

![Screenshot_11](https://user-images.githubusercontent.com/32484496/75624999-12762a00-5bba-11ea-8441-b72f3dc61f2d.png)

`TryToResolve` all it does is create the TypeRef that is needed and use the `.Resolve()`. In case it is resolved, it would be:

![Screenshot_12](https://user-images.githubusercontent.com/32484496/75625025-5832f280-5bba-11ea-87a9-c050de52b2c4.png)

So suppose I want to use `System.Convert `I call this method and get all the original references of the assembly, for example: 
[// System.Drawing.Primitives, Version = 4.2.1.0, Culture = neutral, PublicKeyToken = b03f5f7f11d50a3a
// System.Runtime, Version = 4.2.2.0, Culture = neutral, PublicKeyToken = b03f5f7f11d50a3a
// System.Runtime.Extensions, Version = 4.2.2.0, Culture = neutral, PublicKeyToken = b03f5f7f11d50a3a](url)

try to solve `System.Convert` in each of them, and get it in `System.Runtime.Extensions`, so it returns the resolved type.

**If I add the method solved directly, I get the error that it is in another module.**

![Screenshot_13](https://user-images.githubusercontent.com/32484496/75625125-4ef65580-5bbb-11ea-87c5-463832ea684c.png)

So, I import it this way, everything works perfectly, but the reference `System.Runtime.Extensions` is now twice.

I could understand that this twice if it were different, for example, that I import it manually, but the reference is being obtained from the AssemblyRef of the module, it cannot be that it is another version basically because it obtains the one it already has, that is, it is identical

So why double if it is the same?

I am doing something wrong ? Should I "import" in another way, or is the error "resolving"?

Thank you very much, and sorry to waste your time,

A cordial greeting,

**Comment by juanal98** [comment]
I think the problem may be with the importer,

https://github.com/0xd4d/dnlib/blob/38df36da09832ff870647e9bb451dbc51e97e9b4/src/DotNet/Importer.cs#L251

Maybe I am wrong, but could it be that the importer tries to solve the AssemblyRef and how can it not create a new one using  `CreateScopeReference`?

https://github.com/0xd4d/dnlib/blob/38df36da09832ff870647e9bb451dbc51e97e9b4/src/DotNet/Importer.cs#L350

If I am not wrong, this will work as long as the type is inside the original module or its `System.Runtime / Corlib`, but whenever it is an external library (like this case) it will double the reference.

Do you think it is for this? What would you recommend me?

**Although I am somewhat confused, of course, assuming that the reference is not in the assembly, its operation is correct, because it must add it, perhaps a check in the importer is missing so that you do not add references that are already there?**

**Comment by 0xd4d** [comment]
Something else is different, eg. the assembly flags or hash value.

**Comment by juanal98** [comment]
But how can it be different, if it is being obtained from the references of the application?

**Comment by juanal98** [comment]
As you say, it is different.

When I get:

`var resolvedType = new TypeRefUser(module, @namespace, name, resolutionScope).Resolve();`

The `resolvedType.DefinitionAssembly` and the `resolutionScope` contain different values, but:

**Why does this happen ?**

It should be the same, maybe it's because the `.Resolve()` makes a copy so its `DefinitionAssembly ` is like a copy?

**What solution did you think should be applied?** Because using the `import(resolvedType)`, as it is different, add the reference again (although it is the same).

**Thank you very much for your patience and understanding!**

**Comment by juanal98** [comment]
Hello again,

I have found a solution, but I don't quite reason why.

When using `.Resolve()`  the `TypeDef` that returns does not contain the same `DefinitionAssembly`, as the source (in this case, `resolutionScope`).

![Screenshot_14](https://user-images.githubusercontent.com/32484496/75722868-590f7580-5cdb-11ea-9d24-3de07da9ebc7.png)

The solution has been, once I have validated that it can be resolved,  I have created a `TypeRefUser` again with the `resolutionScope`, and import it.

or what is the same, directly import the `TypeRefUser` and **not** the `TypeDef` that returns the `.Resolve();`

![Screenshot_15](https://user-images.githubusercontent.com/32484496/75723110-efdc3200-5cdb-11ea-9600-f32fed2d1f5b.png)

This would be the example visually, instead of using the `TypeDef`, I create a new `TypeRef` with the original `AssemblyRef`.

This works correctly, now it no longer duplicates the reference, but, this is well done, is it the right way to do it? Or there is some function equivalent to `.Resolve()` but that returns the type correctly.
